src/VisualStudio/Core/Def/Implementation/EditAndContinue/VisualStudioDebugStateChangeListener.cs

```diff
@@ -5,80 +5,194 @@
 using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.EditAndContinue;
 using Microsoft.CodeAnalysis.Editor.Implementation.EditAndContinue;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
-using Microsoft.VisualStudio.Debugger.UI.Interfaces;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.VisualStudio.Debugger.Clr;
 using Roslyn.Utilities;
 
+using Dbg = Microsoft.VisualStudio.Debugger.UI.Interfaces;
+
 namespace Microsoft.VisualStudio.LanguageServices.EditAndContinue
 {
-    [Export(typeof(IDebugStateChangeListener))]
+    [Export(typeof(Dbg.IDebugStateChangeListener))]
     [ExportMetadata("UIContext", Guids.EncCapableProjectExistsInWorkspaceUIContextString)]
-    internal sealed class VisualStudioDebugStateChangeListener : IDebugStateChangeListener
+    internal sealed class VisualStudioDebugStateChangeListener : Dbg.IDebugStateChangeListener
     {
-        private readonly Workspace _workspace;
+        private readonly RemoteEditAndContinueServiceProxy _proxy;
         private readonly IDebuggingWorkspaceService _debuggingService;
         private readonly IActiveStatementTrackingService _activeStatementTrackingService;
+        private readonly IDiagnosticAnalyzerService _diagnosticService;
+        private readonly EditAndContinueDiagnosticUpdateSource _diagnosticUpdateSource;
+        private readonly DebuggeeModuleMetadataProvider _managedModuleInfoProvider;
+
+        private IDisposable? _editSessionConnection;
 
-        // EnC service or null if EnC is disabled for the debug session.
-        private IEditAndContinueWorkspaceService? _encService;
+        private bool _disabled;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioDebugStateChangeListener(VisualStudioWorkspace workspace)
+        public VisualStudioDebugStateChangeListener(
+            VisualStudioWorkspace workspace,
+            Dbg.IManagedModuleInfoProvider managedModuleInfoProvider,
+            IDiagnosticAnalyzerService diagnosticService,
+            EditAndContinueDiagnosticUpdateSource diagnosticUpdateSource)
         {
-            _workspace = workspace;
+            _proxy = new RemoteEditAndContinueServiceProxy(workspace);
             _debuggingService = workspace.Services.GetRequiredService<IDebuggingWorkspaceService>();
             _activeStatementTrackingService = workspace.Services.GetRequiredService<IActiveStatementTrackingService>();
+            _managedModuleInfoProvider = new DebuggeeModuleMetadataProvider(managedModuleInfoProvider);
+            _diagnosticService = diagnosticService;
+            _diagnosticUpdateSource = diagnosticUpdateSource;
         }
 
+#pragma warning disable VSTHRD102 // TODO: Implement internal logic asynchronously
```

Comment:
> // TODO: Implement internal logic asynchronously [](start = 33, length = 49)

Needs a new API from the debugger.

Reply:
yes.  can you file an issue on that at least?  debugger pauses are something i hate :)

Reply:
The work is already in progress. All these types are getting replaced. This has nothing to do with debugger pauses btw.